Mod all cards in stacks, not just first

davidjb opened the pull request:
This change iterates through the `_cards` property (the array of cards in a HuiStackCard, the base class for horizontal or vertical stacks), only if it exists, to apply the same mods to each sub-card in a stack.  Previously, the mods only applied to the first card in a stack given the `document.querySelector` calls return the first matching element only.

If the given card doesn't have the `_cards` property, the same original behaviour is applied, eg to just the card itself, so this is fully backwards compatible.

thomasloven commented:
That's a really neat solution!
Even better, though; make it recursive.

Keep the original `_cardMod` function, but give it a parameter `root`. Then make all calls to `this._cardMod()` be `this._cardMod(this.card)` and just add

´´´javascript
if(root._cards && root._cards.length) {
  root._cards.forEach(c => this._cardMod(c));
  return;
}
```

That way it will work for stacks in stacks as well.
You do that, and I'll change [`layout-card`](https://github.com/thomasloven/lovelace-layout-card) use `_cards` as well. OK?

davidjb commented:
Okay, done!  I made one extra change to support Stacks inside nested CardModders.  My previous code was only looking at whether the current card was a Stack when deciding if to recur; now it also checks if the current card is a CardModder _and_ it contains a Stack.  See  https://github.com/thomasloven/lovelace-card-modder/pull/32/files#diff-5e970dbe79e3a15103e8f0e9a673f761R63

Added a small note to the README as well.

thomasloven commented:
Good idea. I have an idea for optimization, though. Replace all `this.card` (except on line 31) with `this._cards[0]`... and I'm sure you see where I'm going with this...

davidjb committed:
Modify all cards in stacks

This change iterates through the `_cards` property (the array of cards
in a HuiStackCard), only if it exists, to apply the same mods to each
sub-card in a stack.  Previously, the mods only applied to the first
card in a stack given the `document.querySelector` calls return the
first matching element only.

If the given card isn't a Stack or a CardModder containing a Stack, the
same original behaviour is applied, eg to just the card itself.

davidjb commented:
Not sure I follow -- `this.card` is the modified or soon-to-be-modified card, but `this._cards` doesn't exist as the object `this` is an instance of the CardModder class.

(Rebased this PR after the merging of the other one)

thomasloven committed:
Use _cards recursion internally too.

thomasloven commented:
Actually, something like this is even nicer. Is that ok with you?

I realized, btw, that if a style is applied to a stack with a template, only the last card will be updated when the template value changes... Will have to fix that somehow. But that's for later.

davidjb commented:
Aha, gotcha - I see what you meant; very elegant!  All good to merge?

Thanks for the heads up about the stack/template combination.  It's not something I've done (yet) but I'm sure I would have reached that point at some stage.

thomasloven committed:
Add note to readme about templates and stacks

thomasloven commented:
Thanks a lot!
